aten/src/ATen/templates/TensorBody.h

```diff
@@ -494,7 +494,25 @@ class TORCH_API Tensor {
   }
 
   template <typename T>
-  T * data_ptr() const;
+  T * data_ptr() const &;
+
+#define TORCH_DATA_PTR_WARNING                                             \
+  "Calling data_ptr() on a temporary is potentially unsafe and can "       \
+  "indicate use-after-free bug if it was the last reference to the tensor" \
+  " (as in `t.contiguous().data_ptr()`). Please store the Tensor result "  \
+  "in a variable before accessing the data pointer, e.g. "                 \
+  "`Tensor t_contig = t.contiguous(); t_config.data_ptr()`."               \
```

Comment:
we should ask people to use expect_contiguous() in new code IMO, but that's debatable